src/subs.jl

```
@@ -9,23 +9,63 @@ end
+    #   two individual rules (which we don't know of in this method), `z` will not be replaced.
+    if s.first.kind == REAL
+        for j in eachindex(vars)
+            if vars[j] == s.first
+                vals[j] = s.second
+                C == Commutative && break
+            end
+        end
+    else
+        for j in eachindex(vars)
+            if vars[j].variable_order.order.id ==
+               s.first.variable_order.order.id
+                if s.first.kind == COMPLEX || s.first.kind == CONJ
+                    value = s.first.kind == CONJ ? conj(s.second) : s.second
+                    if vars[j].kind == COMPLEX
+                        vals[j] = value
+                    elseif vars[j].kind == CONJ
+                        vals[j] = conj(value)
+                    elseif vars[j].kind == REAL_PART
+                        vals[j] = real(value)
+                    else
+                        vals[j] = imag(value)
+                    end
+                elseif s.first.kind == REAL_PART
+                    isreal(s.second) || error(
+                        "Cannot assign a complex value to the real part of an expression",
+                    )
+                    value = real(s.second) # just to make sure the type is correct
+                    if vars[j].kind == REAL_PART
+                        vals[j] = value
+                    end
+                    # don't do a partial substitution
+                else
+                    @assert(s.first.kind == IMAG_PART)
+                    isreal(s.second) || error(
+                        "Cannot assign a complex value to the imaginary part of an expression",
+                    )
+                    value = real(s.second) # just to make sure the type is correct
+                    if vars[j].kind == IMAG_PART
+                        vals[j] = value
+                    end
+                    # don't do a partial substitution
```

I think we should error here. As you said in your comment:

> I guess that a typical situation will probably never mix real/imaginary part variables with full complex variables, so I don't expect whatever behavior is defined here to actually be used somewhere.

I agree, for this reason, an error is fine because when this is used, this will usually be a mistake in which case an error is helpful. On the other hand, if we silently ignore here, then it will not be helpful debugging these mistakes.

---

Ah, understood. I thought that the previous change is well-defined behavior, but of course it is not documented how substition should work in these cases in the MultivariatePolynomials interface. Now it is an error.